### PR TITLE
Adding login and logout as starchart user from initial seed

### LINKED ISSUE #22: Login / Sign Up - Implementation
A sign up page that authenticates using Seneca email when creating an account.

- I suggest having the user be authenticated for their email instead of having them sign up directly using their Seneca account
- - Complication with Microsoft might occur in the future, and fixing issues can become complicated if it's external
- User display name (if required) can be taken (if possible) directly when the user authorizes the account creation through outlook

A login page that requires the user's email and password to access the web page.

- User should be allowed to have their login creds be saved or not
- Requires user's Seneca email and (custom password) to login to the site

### PR DESCRIPTION
**Pull request by sfrunza13**
Closes #22
Adding login page with login button that signs you (creates a session as per the session server that came with blues stack) in as the user that we seed username:"starchartdev".

### CODE REVIEW

File: app/routes/index.tsx [hunk 1]
```diff
@@ -1,8 +1,26 @@
+import { json, LoaderArgs } from "@remix-run/node";
+import { useLoaderData } from "@remix-run/react";
+import { requireUsername } from "~/session.server";
+
+export const loader = async ({ request }: LoaderArgs) => {
+  const username = await requireUsername(request);
+
+  return json({ username: username });
```
Comment (humphd) [inline app/routes/index.tsx:8]: nit: when the key/value are the same, include it only once:

```ts
return json({ username });
```

File: app/routes/index.tsx [hunk 2]
```diff
@@ -1,8 +1,26 @@
...
+};
+
 export default function Index() {
+  const data = useLoaderData<typeof loader>();
+
   return (
     <main>
       <h1>Starchart</h1>
       <p>Start making your own custom domains and certificates today!</p>
+      <p>{data.username}</p>
```
Comment (humphd) [inline app/routes/index.tsx:18]: Let's use [Chakra UI components](https://chakra-ui.com/docs/components) and [Remix's `<Form>` component](https://remix.run/docs/en/v1/components/form).  Here's a suggestion:

```tsx
<Text>Welcome {data.username}</Text>
<Form action="/logout" method="post">
  <Button type="submit" colorScheme="red">Logout</button>
</Form>
```

File: app/routes/login.tsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    username: "starchartdev",
+    remember: false,
+    redirectTo: "/",
+  });
+};
+
+export default function Login() {
+  return (
+    <Form method="post">
+      <h1>Login</h1>
```
Comment (humphd) [inline app/routes/login.tsx:17]:
```tsx
<Button type="submit" colorScheme="red">Login</button>
```

**Review by humphd — CHANGES_REQUESTED** [review]
This is looking good.  A few things to use newer code, and also you'll need to rebase to get the changes to `index.tsx`

**Comment by sfrunza13** [comment]
> This is looking good. A few things to use newer code, and also you'll need to rebase to get the changes to `index.tsx`

How's it now?

File: app/routes/index.tsx [hunk 4]
```diff
@@ -1,6 +1,17 @@
-import { Heading, Text } from "@chakra-ui/react";
+import { Heading, Text, Button } from "@chakra-ui/react";
+import { json, LoaderArgs } from "@remix-run/node";
```
Comment (humphd) [inline app/routes/index.tsx:2]: `LoaderArgs` is being imported as a type, so you separate it from your other imports and use `import type`.  You do it like this:

```ts
import { json } from "@remix-run/node";

import type { LoaderArgs } from "@remix-run/node";
```

File: app/routes/index.tsx [hunk 5]
```diff
@@ -9,6 +20,12 @@ export default function Index() {
       <Text fontSize="xl">
         Start making your own custom domains and certificates today!
       </Text>
+      <Text>Welcome {data.username}</Text>
+      <form action="/logout" method="post">
```
Comment (humphd) [inline app/routes/index.tsx:24]: `<form>` is HTML, you want `<Form>` from Remix.

File: app/routes/login.tsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import { Heading, Text, Button } from "@chakra-ui/react";
+import { Form } from "@remix-run/react";
+import { ActionArgs } from "@remix-run/node";
```
Comment (humphd) [inline app/routes/login.tsx:3]: Same with this: move to a new line, `import type ...`

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by sfrunza13** [comment]
Hopefully that does it

File: app/routes/index.tsx [hunk 7]
```diff
@@ -1,6 +1,18 @@
-import { Heading, Text } from "@chakra-ui/react";
+import { Heading, Text, Button } from "@chakra-ui/react";
+import { json } from "@remix-run/node";
+import type { LoaderArgs } from "@remix-run/node";
```
Comment (humphd) [inline app/routes/index.tsx:3]: Move your type imports below your regular imports and separate with a blank line:

```ts
import {...} from "..."
import {...} from "..."

import type {...} from "..."
import type {...} from "..."
```

File: app/routes/login.tsx [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import { Heading, Text, Button } from "@chakra-ui/react";
+import { Form } from "@remix-run/react";
+import type { ActionArgs } from "@remix-run/node";
```
Comment (humphd) [inline app/routes/login.tsx:3]: Same thing here.

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by sfrunza13** [comment]
I put the types under other imports and squashed, unfortunately realized I had not ordered things right in the logout and added another commit to fix the order of the imports there too

**Comment by humphd** [comment]
You have conflicts in `app/routes/index.tsx`, you need to rebase and fix.

**Comment by sfrunza13** [comment]
> You have conflicts in `app/routes/index.tsx`, you need to rebase and fix.

is it correct to do as I am doing and pull upstream into origin main, then checkout my feature branch and rebase main and force push?

**Comment by humphd** [comment]
```
git checkout main
git pull upstream main
git checkout your-branch
git rebase main
git push origin your-branch -f
```

**Comment by sfrunza13** [comment]
Cool, that's what I did :D

**Review by humphd — APPROVED** [review]

**Comment by sfrunza13** [comment]
and after all that's done and changes are approved is it ok to just hit squash and merge here in the PR?

**Comment by humphd** [comment]
Not sure why Prettier is failing here, but your code changes look good.

**Comment by sfrunza13** [comment]
> Not sure why Prettier is failing here, but your code changes look good.

this has made me realize there was a format script to rewrite with prettier before my last rebase and now there isn't

**Comment by Eakam1007** [comment]
Looks good. Before merging, you can use `npx prettier --write .` to rewrite with prettier

**Review by Eakam1007 — APPROVED** [review]
Looks good! 
Is there an issue to add end-to-end tests for this once #85 is merged? This page will likely change in the future but I think we should start adding tests early

**Comment by sfrunza13** [comment]
> Looks good! Is there an issue to add end-to-end tests for these pages once #85 is merged? This page will likely change in the future but I think we should start adding tests early

I'll have to learn how to write one but I think I could probably do it :)

**Comment by Eakam1007** [comment]
> I'll have to learn how to write one but I think I could probably do it :)

That's perfect! Do you wanna create an issue for the tests?

**Comment by sfrunza13** [comment]
> > I'll have to learn how to write one but I think I could probably do it :)
> 
> That's perfect! Do you wanna create an issue for the tests?

not super descriptive but #98

**Review by humphd — APPROVED** [review]
Nice work, this looks great.  I'm excited to see the auth happening so soon!

**Pull request merged**